File: src/main/java/org/codetome/hexameter/core/internal/impl/HexagonalGridCalculatorImpl.java
package org.codetome.hexameter.core.internal.impl;

import org.codetome.hexameter.core.api.CubeCoordinate;
import org.codetome.hexameter.core.api.Hexagon;
import org.codetome.hexameter.core.api.HexagonalGrid;
import org.codetome.hexameter.core.api.HexagonalGridCalculator;
import org.codetome.hexameter.core.api.RotationDirection;
import org.codetome.hexameter.core.api.contract.SatelliteData;
import org.codetome.hexameter.core.backport.Optional;

import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;

import static java.lang.Math.abs;
import static java.lang.Math.max;
import static java.lang.Math.min;
import static java.lang.Math.round;

public final class HexagonalGridCalculatorImpl<T extends SatelliteData> implements HexagonalGridCalculator<T> {

    private final HexagonalGrid<T> hexagonalGrid;

    public HexagonalGridCalculatorImpl(final HexagonalGrid hexagonalGrid) {
        this.hexagonalGrid = hexagonalGrid;
    }

    @Override
    public int calculateDistanceBetween(final Hexagon hex0, final Hexagon hex1) {
        final double absX = abs(hex0.getGridX() - hex1.getGridX());
        final double absY = abs(hex0.getGridY() - hex1.getGridY());
        final double absZ = abs(hex0.getGridZ() - hex1.getGridZ());
        return (int) max(max(absX, absY), absZ);
    }

    @Override
    public Set<Hexagon<T>> calculateMovementRangeFrom(final Hexagon<T> hexagon, final int distance) {
        final Set<Hexagon<T>> ret = new HashSet<>();
        for (int x = -distance; x <= distance; x++) {
            for (int y = max(-distance, -x - distance); y <= min(distance, -x + distance); y++) {
                final int z = -x - y;
                final int tmpX = hexagon.getGridX() + x;
                final int tmpZ = hexagon.getGridZ() + z;
                final CubeCoordinate tempCoordinate = CubeCoordinate.fromCoordinates(tmpX, tmpZ);
                if (hexagonalGrid.containsCubeCoordinate(tempCoordinate)) {
                    final Hexagon<T> hex = hexagonalGrid.getByCubeCoordinate(tempCoordinate).get();
                    ret.add(hex);
                }
            }
        }
        return ret;
    }

    @Override
    public Optional<Hexagon<T>> rotateHexagon(final Hexagon<T> originalHex, final Hexagon<T> targetHex, final RotationDirection rotationDirection) {
        final int diffX = targetHex.getGridX() - originalHex.getGridX();
        final int diffZ = targetHex.getGridZ() - originalHex.getGridZ();
        final CubeCoordinate diffCoord = CubeCoordinate.fromCoordinates(diffX, diffZ);
        final CubeCoordinate rotatedCoord = rotationDirection.calculateRotation(diffCoord);
        final CubeCoordinate resultCoord = CubeCoordinate.fromCoordinates(
                originalHex.getGridX() + rotatedCoord.getGridX(),
                originalHex.getGridZ() + rotatedCoord.getGridZ()); // 0, x,
        return hexagonalGrid.getByCubeCoordinate(resultCoord);
    }

    @Override
    public Set<Hexagon<T>> calculateRingFrom(final Hexagon<T> centerHexagon, final int radius) {
        final Set<Hexagon<T>> result = new HashSet<>();
        final int neighborIndex = 0;
        Hexagon<T> currentHexagon = centerHexagon;
        for (int i = 0; i < radius; i++) {
            final Optional<Hexagon<T>> neighbor = hexagonalGrid.getNeighborByIndex(currentHexagon, neighborIndex);
            if (neighbor.isPresent()) {
                currentHexagon = neighbor.get();
            } else {
                return result;
            }
        }
        return result;
    }

    @Override
    public List<Hexagon<T>> drawLine(final Hexagon<T> from, final Hexagon<T> to) {
        final int distance = calculateDistanceBetween(from, to);
        final List<Hexagon<T>> results = new LinkedList<>();
        if (distance == 0) {
            return results;
        }
        for (int i = 0; i <= distance; i++) {
            final CubeCoordinate interpolatedCoordinate = cubeLinearInterpolate(from.getCubeCoordinate(),
                    to.getCubeCoordinate(), (1.0 / distance) * i);
            results.add(hexagonalGrid.getByCubeCoordinate(interpolatedCoordinate).get());
        }
        return results;
    }

    @Override
    public boolean isVisible(final Hexagon<T> from, final Hexagon<T> to) {
        final List<Hexagon<T>> traversePath = drawLine(from, to);
        for (Hexagon<T> pathHexagon : traversePath) {
            if (pathHexagon.equals(from) || pathHexagon.equals(to)) {
                continue;
            }
            if (pathHexagon.getSatelliteData().isPresent() && pathHexagon.getSatelliteData().get().isOpaque()) {
                return false;
            }
        }
        return true;
    }

    private CubeCoordinate cubeLinearInterpolate(final CubeCoordinate from, final CubeCoordinate to, final double sample) {
        return roundToCubeCoordinate(linearInterpolate(from.getGridX(), to.getGridX(), sample),
                linearInterpolate(from.getGridY(), to.getGridY(), sample),
                linearInterpolate(from.getGridZ(), to.getGridZ(), sample));
    }

    private double linearInterpolate(final int from, final int to, final double sample) {
        return from + (to - from) * sample;
    }

    private CubeCoordinate roundToCubeCoordinate(final double gridX, final double gridY, final double gridZ) {
        int rx = (int) round(gridX);
        final int ry = (int) round(gridY);
        int rz = (int) round(gridZ);

        final double differenceX = abs(rx - gridX);
        final double differenceY = abs(ry - gridY);
        final double differenceZ = abs(rz - gridZ);

        if (differenceX > differenceY && differenceX > differenceZ) {
            rx = -ry - rz;
        } else if (differenceY <= differenceZ) {
            rz = -rx - ry;
        }
        return CubeCoordinate.fromCoordinates(rx, rz);
    }
}

File: mixite.core/core/src/main/kotlin/org/hexworks/mixite/core/internal/impl/HexagonalGridCalculatorImpl.kt
package org.hexworks.mixite.core.internal.impl

import org.hexworks.mixite.core.api.*
import org.hexworks.mixite.core.api.contract.SatelliteData
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min
import kotlin.math.round

class HexagonalGridCalculatorImpl<T : SatelliteData>(private val hexagonalGrid: HexagonalGrid<T>) : HexagonalGridCalculator<T> {

    override fun calculateDistanceBetween(hex0: Hexagon<T>, hex1: Hexagon<T>): Int {
        val absX = abs(hex0.gridX - hex1.gridX)
        val absY = abs(hex0.gridY - hex1.gridY)
        val absZ = abs(hex0.gridZ - hex1.gridZ)
        return max(max(absX, absY), absZ)
    }

    override fun calculateMovementRangeFrom(hexagon: Hexagon<T>, distance: Int): Set<Hexagon<T>> {
        val ret = HashSet<Hexagon<T>>()
        for (x in -distance..distance) {
            for (y in max(-distance, -x - distance)..min(distance, -x + distance)) {
                val z = -x - y
                val tmpX = hexagon.gridX + x
                val tmpZ = hexagon.gridZ + z
                val tempCoordinate = CubeCoordinate.fromCoordinates(tmpX, tmpZ)
                if (hexagonalGrid.containsCubeCoordinate(tempCoordinate)) {
                    val hex = hexagonalGrid.getByCubeCoordinate(tempCoordinate).get()
                    ret.add(hex)
                }
            }
        }
        return ret
    }

    override fun rotateHexagon(originalHex: Hexagon<T>, targetHex: Hexagon<T>, rotationDirection: RotationDirection): Maybe<Hexagon<T>> {
        val diffX = targetHex.gridX - originalHex.gridX
        val diffZ = targetHex.gridZ - originalHex.gridZ
        val diffCoord = CubeCoordinate.fromCoordinates(diffX, diffZ)
        val rotatedCoord = rotationDirection.calculateRotation(diffCoord)
        val resultCoord = CubeCoordinate.fromCoordinates(
                originalHex.gridX + rotatedCoord.gridX,
                originalHex.gridZ + rotatedCoord.gridZ) // 0, x,
        return hexagonalGrid.getByCubeCoordinate(resultCoord)
    }

    override fun calculateRingFrom(centerHexagon: Hexagon<T>, radius: Int): Set<Hexagon<T>> {
        val result = HashSet<Hexagon<T>>()
        val neighborIndex = 0
        var currentHexagon = centerHexagon
        for (i in 0 until radius) {
            val neighbor = hexagonalGrid.getNeighborByIndex(currentHexagon, neighborIndex)
            if (neighbor.isPresent) {
                currentHexagon = neighbor.get()
            } else {
                return result
            }
        }
        return result
    }

    override fun drawLine(from: Hexagon<T>, to: Hexagon<T>): List<Hexagon<T>> {
        val distance = calculateDistanceBetween(from, to)
        val results = mutableListOf<Hexagon<T>>()
        if (distance == 0) {
            return results
        }
        for (i in 0..distance) {
            val interpolatedCoordinate = cubeLinearInterpolate(from.cubeCoordinate,
                    to.cubeCoordinate, 1.0 / distance * i)
            results.add(hexagonalGrid.getByCubeCoordinate(interpolatedCoordinate).get())
        }
        return results
    }

    override fun isVisible(from: Hexagon<T>, to: Hexagon<T>): Boolean {
        val traversePath = drawLine(from, to)
        for (pathHexagon in traversePath) {
            if (pathHexagon.equals(from) || pathHexagon.equals(to)) {
                continue
            }
            if (pathHexagon.satelliteData.isPresent && pathHexagon.satelliteData.get().opaque) {
                return false
            }
        }
        return true
    }

    private fun cubeLinearInterpolate(from: CubeCoordinate, to: CubeCoordinate, sample: Double): CubeCoordinate {
        return roundToCubeCoordinate(linearInterpolate(from.gridX, to.gridX, sample),
                linearInterpolate(from.gridY, to.gridY, sample),
                linearInterpolate(from.gridZ, to.gridZ, sample))
    }

    private fun linearInterpolate(from: Int, to: Int, sample: Double): Double {
        return from + (to - from) * sample
    }

    private fun roundToCubeCoordinate(gridX: Double, gridY: Double, gridZ: Double): CubeCoordinate {
        var rx = round(gridX).toInt()
        val ry = round(gridY).toInt()
        var rz = round(gridZ).toInt()

        val differenceX = abs(rx - gridX)
        val differenceY = abs(ry - gridY)
        val differenceZ = abs(rz - gridZ)

        if (differenceX > differenceY && differenceX > differenceZ) {
            rx = -ry - rz
        } else if (differenceY <= differenceZ) {
            rz = -rx - ry
        }
        return CubeCoordinate.fromCoordinates(rx, rz)
    }
}
